test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,83 @@
+package upgrade
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+
+	"github.com/iancoleman/strcase"
+	v1 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v1"
+	"github.com/infinispan/infinispan-operator/pkg/controller/constants"
+	ispnctrl "github.com/infinispan/infinispan-operator/pkg/controller/infinispan"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	CrdPath = "../../../deploy/crds/"
+)
+
+var (
+	testKube         = tutils.NewTestKubernetes(os.Getenv("TESTING_CONTEXT"))
+	upgradeStateFlow = []v1.ConditionType{v1.ConditionUpgrade, v1.ConditionStopping, v1.ConditionWellFormed}
+)
+
+// TODO remove and replace with DefaultSpec once Batch PR has been merged as it removes this logic from main_test
+var MinimalSpec = v1.Infinispan{
+	TypeMeta: tutils.InfinispanTypeMeta,
+	ObjectMeta: metav1.ObjectMeta{
+		Name: tutils.DefaultClusterName,
+	},
+	Spec: v1.InfinispanSpec{
+		Replicas: 1,
+	},
+}
+
+func TestGracefulShutdown(t *testing.T) {
+	namespace := tutils.Namespace
+	testKube.NewNamespace(namespace)
+	// Utilise the sha256 of the current Infinispan image to trick the operator into thinking a upgrade is required
+	oldImage := getDockerImageSha()
+	tutils.ExpectNoError(os.Setenv("DEFAULT_IMAGE", oldImage))
+	stopCh := testKube.InstallAndRunOperator(namespace, CrdPath, false)
```

Comment:
@ryanemerson I can't understand how this could help with real operator upgrade. The operator upgrade idea is that initial infra (Pods, Services, Routes, etc.) and CR's are forming with one operator version and after upgrade become to continue with process inside another one.

Reply:
Oh, I see the case in the description, this is not full life-cycle test, but just an upgrade inside one operator version.

Reply:
This allows us to verify that, upon detecting that an upgrade is required, the correct actions are taken by the Operator. In the case of GracefulShutdown this requires scaling down/up the StatefulSet. When #658 is implemented, we will need to create a similar test to ensure that zero-downtime migration occurs.

These tests are a mock of an upgrade and allow for us to ensure that a given PR does not break the upgrade workflow. It does not protect against version incompatibilities and is intended to supplement, rather than replace, the existing manual tests.